db/version_set.cc

```diff
@@ -4881,62 +4881,111 @@ Status VersionSet::WriteSnapshot(log::Writer* log) {
+      total_full_size +=
+          ApproximateSizeLevel0(v, files_brief, start, end, caller);
       continue;
     }
 
     assert(level > 0);
     assert(files_brief.num_files > 0);
 
-    // identify the file position for starting key
-    const uint64_t idx_start = FindFileInRange(
-        v->cfd_->internal_comparator(), files_brief, start,
-        /*start=*/0, static_cast<uint32_t>(files_brief.num_files - 1));
-    assert(idx_start < files_brief.num_files);
-
-    // scan all files from the starting position until the ending position
-    // inferred from the sorted order
-    for (uint64_t i = idx_start; i < files_brief.num_files; i++) {
-      uint64_t val;
-      val = ApproximateSize(v, files_brief.files[i], end, caller);
-      if (!val) {
-        // the files after this will not have the range
-        break;
-      }
+    // identify the file position for start key
+    const int idx_start =
+        FindFileInRange(icmp, files_brief, start, 0,
+                        static_cast<uint32_t>(files_brief.num_files - 1));
+    assert(static_cast<size_t>(idx_start) < files_brief.num_files);
+
+    // identify the file position for end key
+    int idx_end = idx_start;
+    if (icmp.Compare(files_brief.files[idx_end].largest_key, end) < 0) {
+      idx_end =
+          FindFileInRange(icmp, files_brief, end, idx_start,
+                          static_cast<uint32_t>(files_brief.num_files - 1));
+    }
+    assert(idx_end >= idx_start &&
+           static_cast<size_t>(idx_end) < files_brief.num_files);
+
+    // scan all files from the starting index to the ending index
+    // (inferred from the sorted order)
+
+    // first scan all the intermediate full files (excluding first and last)
+    for (int i = idx_start + 1; i < idx_end; ++i) {
+      uint64_t file_size = files_brief.files[i].fd.GetFileSize();
+      // The entire file falls into the range, so we can just take its size.
+      assert(file_size ==
+             ApproximateSize(v, files_brief.files[i], end, caller));
+      total_full_size += file_size;
+    }
+
+    // save the first and the last files (which may be the same file), so we
+    // can scan them later.
+    first_files.push_back(&files_brief.files[idx_start]);
+    if (idx_start != idx_end) {
+      // we need to estimate size for both files, only if they are different
+      last_files.push_back(&files_brief.files[idx_end]);
+    }
+  }
+
+  // Now scan all the first & last files at each level, and estimate their size.
+  uint64_t total_partial_size = 0;
+
+  // Estimate for all the first files, at each level
+  for (const auto file_ptr : first_files) {
+    uint64_t file_size = file_ptr->fd.GetFileSize();
+    if (total_full_size / 10 > file_size) {
+      // The file size is less than 10% of the total_full_size - approximate the
+      // result to avoid the costly binary search inside ApproximateSize.
+      total_partial_size += file_size / 2;
+      continue;
+    }
```

Comment:
Do you still plan to use the total partial file size, rather than single file size?

Reply:
sorry, didn't get the question.

total_full_size counts the full files + L0
total_partial_size counts the partial sizes.

both total_partial_size and total_full_size sum over all the files.

Reply:
This is what I suggest:
```
uint64_t total_boundary_file_size = 0;
for (const auto* file_ptr : first_files) {
   total_boundary_file_size += file_ptr->fd.GetFileSize();
}
for (const auto* file_ptr : last_files) {
   total_boundary_file_size += file_ptr->fd.GetFileSize();
}
if (total_boundary_file_size < total_full_size / 10) {
  total_full_size += total_boundary_file_size / 2;
} else {
  // search into every file.
}
```

Reply:
That is doable. But why not do it on a per file basis?
IIUC, the size of the files at each next level is growing exponentially.
The way, I implemented the logic - we would do binary search on higher levels, but skip doing it on lower levels (closer to L0). So it's not all-or-nothing, but rather more "adaptive" based on the file sizes at each level.
It's of course your call at the end of the day. I am just trying to understand the rationale behind your proposal.